FnCAS/README.md

```diff
@@ -1,8 +1,131 @@
+template <typename T>
+T simple_function(const std::vector<T>& x) {
+  ++number_of_calls;
+  assert(x.size() == 2u);
+  // `sqr(x)` is a convenience defined within `fncas::` for ML purposes,
+  // along with `ramp(v)` and `unit_step(v)`.
+  return fncas::sqr(x[0] + 1) + fncas::sqr(x[1] + 2);
+  // Alternatively, `#define INJECT_FNCAS_INTO_NAMESPACE_STD` and use
+  // the mathematical library from within the `std::` namespace.
+}
+
+// Make a few native C++ calls.
+number_of_calls = 0;
+EXPECT_EQ(5, simple_function(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, simple_function(std::vector<double>({-5, -5})));
+ASSERT_EQ(2, number_of_calls);
+
+// Wrap a native C++ function into an FnCAS type.
+// This type allows passing functions around without thinking of whether they
+// are native functions, blueprints, or JIT-compiled dynamically linked `.so`-s.
+// The `2` parameter is the dimensionality of the function.
+fncas::function_reference_t native(simple_function<double>, 2);
+const fncas::function_t& reference = native;
+
+number_of_calls = 0;
+EXPECT_EQ(5, reference(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, reference(std::vector<double>({-5, -5})));
+ASSERT_EQ(2, number_of_calls);  // By-reference evaluation doesn't call the function.
+
+// Create the blueprint of this function: its internal tree representation.
+// The scope of `x` would be where the blueprint and its uses are valid
+// from within this particular thread. It uses a thread-local singleton.
+fncas::variables_vector_t x(2);
+number_of_calls = 0;
+fncas::function_blueprint_t blueprint = simple_function(x);
+ASSERT_EQ(1, number_of_calls);
+number_of_calls = 0;
+EXPECT_EQ(5, blueprint(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, blueprint(std::vector<double>({-5, -5})));
+ASSERT_EQ(0, number_of_calls);  // Blueprint evalution doesn't call the function.
+
+// Internal only: Examine the textual representation of the blueprint.
+// TODO(dkorolev): One day we may want to clean this syntax and its result.
+EXPECT_EQ("(sqr((x[0]+1))+sqr((x[1]+2)))",
+          blueprint.debug_as_string());
+
+// Create the JIT-compiled representation of the function.
+const fncas::function_compiled_t jit(blueprint);
+number_of_calls = 0;
+EXPECT_EQ(5, jit(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, jit(std::vector<double>({-5, -5})));
+ASSERT_EQ(0, number_of_calls);
+
+// Confirm both the blueprint and the JIT version can be cast down to `function_t`.
+const fncas::function_t& reference = blueprint;
+number_of_calls = 0;
+EXPECT_EQ(5, reference(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, reference(std::vector<double>({-5, -5})));
+ASSERT_EQ(0, number_of_calls);
+
+const fncas::function_t& jit_reference = jit;
+number_of_calls = 0;
+EXPECT_EQ(5, jit_reference(std::vector<double>({0, 0})));
+EXPECT_EQ(4*4 + 3*3, jit_reference(std::vector<double>({-5, -5})));
+ASSERT_EQ(0, number_of_calls);
+
+// Wrap the function into the approximate gradient computer.
```

Comment:
Would be good to explain in a couple of words, what does `approximate gradient` mean.